dpctl/tensor/libtensor/include/kernels/accumulators.hpp

```diff
@@ -66,184 +68,312 @@ template <typename inputT, typename outputT> struct NonZeroIndicator
+    acc_groups = ceiling_quotient<size_t>(acc_nelems, n_wi * wg_size);
 
     sycl::event inc_scan_phase1_ev = exec_q.submit([&](sycl::handler &cgh) {
         cgh.depends_on(depends);
 
         using slmT = sycl::local_accessor<outputT, 1>;
 
+        auto gws = sycl::range<1>(iter_nelems * acc_groups * wg_size);
         auto lws = sycl::range<1>(wg_size);
-        auto gws = sycl::range<1>(n_groups * wg_size);
 
         auto ndRange = sycl::nd_range<1>(gws, lws);
 
         slmT slm_iscan_tmp(lws, cgh);
 
-        using KernelName =
-            inclusive_scan_iter_local_scan_krn<inputT, outputT, n_wi, IndexerT,
-                                               TransformerT>;
+        using KernelName = inclusive_scan_iter_local_scan_krn<
+            inputT, outputT, n_wi, IterIndexerT, InpIndexerT, OutIndexerT,
+            TransformerT, ScanOpT, include_initial>;
 
         cgh.parallel_for<KernelName>(ndRange, [=, slm_iscan_tmp =
                                                       std::move(slm_iscan_tmp)](
                                                   sycl::nd_item<1> it) {
-            auto chunk_gid = it.get_global_id(0);
-            auto lid = it.get_local_id(0);
+            size_t gid = it.get_global_id(0);
+            size_t lid = it.get_local_id(0);
+
+            size_t iter_gid = gid / (acc_groups * wg_size);
+            size_t chunk_gid = gid - (iter_gid * acc_groups * wg_size);
 
-            std::array<outputT, n_wi> local_isum;
+            std::array<outputT, n_wi> local_iscan;
 
             size_t i = chunk_gid * n_wi;
+            auto iter_offsets = iter_indexer(iter_gid);
+            const auto &inp_iter_offset = iter_offsets.get_first_offset();
+            const auto &out_iter_offset = iter_offsets.get_second_offset();
 
 #pragma unroll
             for (nwiT m_wi = 0; m_wi < n_wi; ++m_wi) {
-                constexpr outputT out_zero(0);
-
-                local_isum[m_wi] =
-                    (i + m_wi < n_elems)
-                        ? transformer(input[indexer(s0 + s1 * (i + m_wi))])
-                        : out_zero;
+                if constexpr (!include_initial) {
+                    local_iscan[m_wi] =
+                        (i + m_wi < acc_nelems)
+                            ? transformer(
+                                  input[inp_iter_offset +
+                                        inp_indexer(s0 + s1 * (i + m_wi))])
+                            : identity;
+                }
+                else {
+                    // shift input to the left by a single element relative to
+                    // output
+                    local_iscan[m_wi] =
+                        (i + m_wi < acc_nelems && i + m_wi > 0)
+                            ? transformer(
+                                  input[inp_iter_offset +
+                                        inp_indexer((s0 + s1 * (i + m_wi)) -
+                                                    1)])
```

Comment:
Can `s0` and `s1` ever be zeros?

Reply:
No, it shouldn't be possible, and especially not where `include_initial` is true.

They're first set (as constexpr) to 0 and 1 respectively, then later, `chunk_size - 1` and `chunk_size` if more than 1 group is necessary. Since `chunk_size` is `wg_size * n_wi`, they can never both be 0.